### PR TITLE
Gracefully handle GOAWAY

### PR DESCRIPTION
**Pull request by dagnir**
## Description
- Ensure that the GOAWAY exception is fired within each child channel's event
   loop to ensure correct ordering

 - Fix off-by-one error in MultiplexedChannelRecord

 - Add tests using a minimal H2 endpoint implementation

 - Remove test-server module since it's no longer needed

## Motivation and Context
Bug fix.

## Testing
 - New unit tests
 - Integration tests
 - Will run stability tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        sendGetRequest().join();
+
+        // Note: It's possible the initial request can cause the client to allocate more than 1 channel
+        int initialChannelNum = endpointDriver.channels.size();
+
+        // Send GOAWAY to all the currently open channels
+        endpointDriver.channels.forEach(ch -> endpointDriver.goAway(ch, 1));
+
+        // Need to give a chance for the streams to get closed
+        Thread.sleep(1000);
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java]: Have we run this enough to be convinced it isn't flaky?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+                if (streamId < 5) {
+                    Http2Headers outboundHeaders = new DefaultHttp2Headers()
+                            .status("200")
+                            .add("content-type", "text/plain")
+                            .addInt("content-length", getPayload.length);
+
+                    frameWriter().writeHeaders(ctx, streamId, outboundHeaders, 0, false, ctx.newPromise());
+                    ctx.flush();
+                }
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java]: I'm probably missing something. Why 3 and 5 when there's only 2 requests?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPoolTest.java [hunk 3]
```diff
@@ -62,7 +74,7 @@ public void closeWaitsForConnectionToBeReleasedBeforeClosingConnectionPool() thr
             Promise<Void> releasePromise = Mockito.spy(new DefaultPromise<>(loopGroup.next()));
             Mockito.doCallRealMethod().when(releasePromise).await();
             releasePromise.setSuccess(null);
-            Mockito.when(connectionPool.release(Mockito.eq(channel))).thenReturn(releasePromise);
+            Mockito.when(connectionPool.release(eq(channel))).thenReturn(releasePromise);
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPoolTest.java]: Do we need this file, still?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java (hunk 1 above)
Comment (dagnir) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java]: 1s is a pretty long time to close 2 connections so it should be fine

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java (hunk 2 above)
Comment (dagnir) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java]: [The client always initiates streams using odd numbers](https://httpwg.org/specs/rfc7540.html#StreamIdentifiers). The first stream used to establish the H2 stream is 1, so the next two are 3 and 5.

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPoolTest.java (hunk 3 above)
Comment (dagnir) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPoolTest.java]: I can revert this

Comment (dagnir) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPoolTest.java]: [](url)

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java (hunk 2 above)
Comment (millems) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java]: Ah, cool!

**Pull request merged**